### PR TITLE
Add user authentication and profile servlet and html page

### PR DESCRIPTION
**Pull request by ALai2**
I added user authentication with login and logout urls and started implementing code for profile. There is a profile.html page which currently only has a form which currently does nothing when the submit button is pressed. I also created a ProfileServlet.java which I will test when the code for the profile javascript is implemented. There is a bunch of other committed files in a target folder because it was added due to me testing the code with the command "mvn package appengine:run".

### CODE REVIEW

File: src/main/webapp/profile.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>Insert cool name here</title>
```
Comment (joeychin01) [inline src/main/webapp/profile.html:5]: Probably should change this

File: src/main/java/com/google/sps/servlets/ProfileServlet.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
```
[the remark this answers is not in the document]
Comment (ALai2) [reply src/main/java/com/google/sps/servlets/ProfileServlet.java:37]: I don't, thanks for pointing it out!

File: src/main/java/com/google/sps/data/Profile.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  // dietary options
+  private boolean vegetarian;
+  private boolean vegan;
+  private boolean glutenFree;
+  private boolean dairyFree;
+
+  private String[] allergies;
+  
+  // profile picture?
```
[the remark this answers is not in the document]
Comment (ALai2) [reply src/main/java/com/google/sps/data/Profile.java:35]: The profile picture is definitely going to be a stretch idea. I forget to take the comment out when I commit my changes.

File: src/main/webapp/profile.html (hunk 1 above)
Comment (ALai2) [reply src/main/webapp/profile.html]: Title has been changed to "Your Profile".

File: src/main/webapp/script.js [hunk 4]
```diff
@@ -10,4 +10,53 @@
...
+    if (userInfo.isLoggedIn) {
+      const logoutElement = document.createElement('a');
+      logoutElement.innerHTML = "Logout";
+      logoutElement.href = userInfo.logoutUrl;
+
+      const textElement = document.createElement('p');
+      if (userInfo.hasProfile) {
+        textElement.innerHTML = "Welcome, <strong>" + userInfo.userName + "</strong>";
+      } else {
+        textElement.innerHTML = "Welcome! Remember to set a profile!";
```
Comment (cocovsack) [inline src/main/webapp/script.js:43]: "create a profile"

File: target/portfolio-1/script.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      loginElement.href = userInfo.loginUrl;
+
+      const textElement = document.createElement('p');
+      textElement.innerHTML = "Hello!";
+
+      loginStatusElement.appendChild(loginElement);
+      loginStatusElement.appendChild(textElement);
+    }
+  });
+
```
Comment (cocovsack) [inline target/portfolio-1/script.js:61]: nit: delete extra line

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.*;
```
[the remark this answers is not in the document]
Comment (ALai2) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: I can be more specific.

File: src/main/java/com/google/sps/data/Profile.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  /**
+    * @param id The unique id of the profile.
+    * @param userName The name of the user that the profile belongs to.
+    * @param allergies The allergies of the user
+    * @param vegetarian @param vegan @param glutenFree @param diaryFree The user's dietary options
+    */
+  public Profile(long id, String userName, boolean vegetarian, boolean vegan, boolean glutenFree, boolean dairyFree, String[] allergies) {
```
Comment (linxiz) [inline src/main/java/com/google/sps/data/Profile.java:43]: Could you use Builder pattern?

File: src/main/java/com/google/sps/data/Profile.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  private String userName;
+  private String userEmail;
+
+  // dietary options
+  private boolean vegetarian;
+  private boolean vegan;
+  private boolean glutenFree;
+  private boolean dairyFree;
+
+  private String[] allergies;
```
Comment (linxiz) [inline src/main/java/com/google/sps/data/Profile.java:33]: Do you plan to allow all raw text values?

File: src/main/java/com/google/sps/servlets/ProfileServlet.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      responseMap.put("error", errorMessage);
+      
+    } else {
+      String id = userService.getCurrentUser().getUserId();
+      String userName = request.getParameter("userName");
+      boolean vegetarian = Boolean.parseBoolean(request.getParameter("vegetarian"));
+      boolean vegan = Boolean.parseBoolean(request.getParameter("vegan"));
+      boolean glutenFree = Boolean.parseBoolean(request.getParameter("glutenFree"));
+      boolean dairyFree = Boolean.parseBoolean(request.getParameter("dairyFree"));
+      String[] allergies = request.getParameterValues("allergies");
```
Comment (linxiz) [inline src/main/java/com/google/sps/servlets/ProfileServlet.java:97]: nit: prefer to be closer to where first used, you could also consider combining them in one line with the entity.setProperty

Comment (linxiz) [inline src/main/java/com/google/sps/servlets/ProfileServlet.java:97]: Do you want to set an upper limit for the array size?

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.json.simple.JSONObject;
+import com.google.gson.Gson;
+
+/** Servlet that deals with login and logout requests */
+@WebServlet("/user")
+public class UserServlet extends HttpServlet {
```
Comment (linxiz) [inline src/main/java/com/google/sps/servlets/UserServlet.java:37]: nit: Consider renaming to login to be more readable, get user sounds a bit like getting profile

File: src/main/java/com/google/sps/servlets/ProfileServlet.java (hunk 9 above)
Comment (linxiz) [inline src/main/java/com/google/sps/servlets/ProfileServlet.java:97]: Could these be null? Could you add tests when some of the parameter is not set in the request? Also, do you intend to allow empty user name?

Comment (ALai2) [reply src/main/java/com/google/sps/servlets/ProfileServlet.java:97]: These could be null. parseBoolean returns false when the parameter is null which is the correct behavior when the input isn't true. The username is required in the html form, I do not intend to allow empty username. I added a check with an error message in the servlet. I also added a check for empty parameter check for String[] allergies.

File: src/main/java/com/google/sps/data/Profile.java (hunk 8 above)
Comment (ALai2) [reply src/main/java/com/google/sps/data/Profile.java:33]: Yes, but later on there might be options for the user to choose from so that random values won't get used.

File: src/main/java/com/google/sps/servlets/ProfileServlet.java (hunk 9 above)
Comment (ALai2) [reply src/main/java/com/google/sps/servlets/ProfileServlet.java]: I could set an upper limit but in case a person has a ton of allergies I don't want to assume that there is an upper bound on the amount of allergies a person can have and remove the extra ones.

File: src/main/java/com/google/sps/data/Profile.java (hunk 7 above)
Comment (ALai2) [reply src/main/java/com/google/sps/data/Profile.java:43]: Yes, I could add a ProfileBuilder class which constructs a profile.

**Comment by ALai2** [comment]
> I add .gitignore to the project, could you revert all your files in /target?

I removed the /target files from my commit.
